## ELT (Extract, Transform, Load) da nossa base

In [54]:
#importando as libs necessarias

import pandas as pd #manipulação e tratamento dos dados
import psycopg2 as pg2 #para realizar a conexão com o Postgrees
from sqlalchemy import create_engine #auxilia na conexão do Postgrees

In [ ]:
#instalando as bibliotecas

pip install pandas 
pip install psycopg2 
pip install sqlalchemy 

In [55]:
#importando nosso dataframe e lendo as primeiras linhas

df = pd.read_csv(r'C:\Users\Ana Carolina Meliti\OneDrive\Área de Trabalho\SalesRetail\CC_GENERAL.csv')
df.head()

CUST_ID      BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0  C10001    40.900749           0.818182      95.40              0.00   
1  C10002  3202.467416           0.909091       0.00              0.00   
2  C10003  2495.148862           1.000000     773.17            773.17   
3  C10004  1666.670542           0.636364    1499.00           1499.00   
4  C10005   817.714335           1.000000      16.00             16.00   

   INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                    95.4      0.000000             0.166667   
1                     0.0   6442.945483             0.000000   
2                     0.0      0.000000             1.000000   
3                     0.0    205.788017             0.083333   
4                     0.0      0.000000             0.083333   

   ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                    0.000000                          0.083333   
1                    0.000000                          0.000000   
2                    1.000000                          0.000000   
3                    0.083333                          0.000000   
4                    0.083333                          0.000000   

   CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                0.000000                 0              2        1000.0   
1                0.250000                 4              0        7000.0   
2                0.000000                 0             12        7500.0   
3                0.083333                 1              1        7500.0   
4                0.000000                 0              1        1200.0   

      PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
0   201.802084        139.509787          0.000000      12  
1  4103.032597       1072.340217          0.222222      12  
2   622.066742        627.284787          0.000000      12  
3     0.000000               NaN          0.000000      12  
4   678.334763        244.791237          0.000000      12

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8950 entries, 0 to 8949
Data columns (total 18 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   CUST_ID                           8950 non-null   object 
 1   BALANCE                           8950 non-null   float64
 2   BALANCE_FREQUENCY                 8950 non-null   float64
 3   PURCHASES                         8950 non-null   float64
 4   ONEOFF_PURCHASES                  8950 non-null   float64
 5   INSTALLMENTS_PURCHASES            8950 non-null   float64
 6   CASH_ADVANCE                      8950 non-null   float64
 7   PURCHASES_FREQUENCY               8950 non-null   float64
 8   ONEOFF_PURCHASES_FREQUENCY        8950 non-null   float64
 9   PURCHASES_INSTALLMENTS_FREQUENCY  8950 non-null   float64
 10  CASH_ADVANCE_FREQUENCY            8950 non-null   float64
 11  CASH_ADVANCE_TRX                  8950 non-null   int64  
 12  PURCHA

Dados Nulos e Duplicados


Nosso dataframe contem 18 colunas com 8949 linhas e não possui dados duplicados, porém há 313 dados nulos na coluna "MINIMUM_PAYMENTS" e 1 dado nulo na coluna "CREDIT_LIMIT", antes de darmos inicio as análises precisamos entender do que se trata os nulos para tratar-los da melhor forma.


A maior parte dos nossos dados são números (tipo inteiros e float).

In [57]:
#Mapeando a quantidade de dados nulos

df.isnull().sum()

CUST_ID                               0
BALANCE                               0
BALANCE_FREQUENCY                     0
PURCHASES                             0
ONEOFF_PURCHASES                      0
INSTALLMENTS_PURCHASES                0
CASH_ADVANCE                          0
PURCHASES_FREQUENCY                   0
ONEOFF_PURCHASES_FREQUENCY            0
PURCHASES_INSTALLMENTS_FREQUENCY      0
CASH_ADVANCE_FREQUENCY                0
CASH_ADVANCE_TRX                      0
PURCHASES_TRX                         0
CREDIT_LIMIT                          1
PAYMENTS                              0
MINIMUM_PAYMENTS                    313
PRC_FULL_PAYMENT                      0
TENURE                                0
dtype: int64

In [58]:
#Mapeando dados duplicados

df.duplicated().sum()

0

Visualizando os valores nulos da coluna 'MINIMUM_PAYMENTS', entendendo antes o que é o pagamento mínimo do cartão de crédito, é o menor valor que pode pagar para não ficar em atraso com o banco. 
Dessa forma optei por trocar o nulo para 0.0 afim de não atrapalhar em nossa análise descritiva posteriormente.

In [59]:
#Visualizando os nulos
nulos = df[df.isnull().any(axis=1)]
nulos

CUST_ID      BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
3     C10004  1666.670542           0.636364    1499.00           1499.00   
45    C10047  2242.311686           1.000000     437.00             97.00   
47    C10049  3910.111237           1.000000       0.00              0.00   
54    C10056     6.660517           0.636364     310.00              0.00   
55    C10057  1311.995984           1.000000    1283.90           1283.90   
...      ...          ...                ...        ...               ...   
8919  C19160    14.524779           0.333333     152.00            152.00   
8929  C19170   371.527312           0.333333       0.00              0.00   
8935  C19176   183.817004           1.000000     465.90              0.00   
8944  C19185   193.571722           0.833333    1012.73           1012.73   
8946  C19187    19.183215           1.000000     300.00              0.00   

      INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
3                        0.0    205.788017             0.083333   
45                     340.0    184.648692             0.333333   
47                       0.0   1980.873201             0.000000   
54                     310.0      0.000000             0.666667   
55                       0.0      0.000000             0.250000   
...                      ...           ...                  ...   
8919                     0.0      0.000000             0.333333   
8929                     0.0   1465.407927             0.000000   
8935                   465.9      0.000000             1.000000   
8944                     0.0      0.000000             0.333333   
8946                   300.0      0.000000             1.000000   

      ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
3                       0.083333                          0.000000   
45                      0.083333                          0.333333   
47                      0.000000                          0.000000   
54                      0.000000                          0.666667   
55                      0.250000                          0.000000   
...                          ...                               ...   
8919                    0.333333                          0.000000   
8929                    0.000000                          0.000000   
8935                    0.000000                          0.833333   
8944                    0.333333                          0.000000   
8946                    0.000000                          0.833333   

      CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
3                   0.083333                 1              1        7500.0   
45                  0.166667                 2              5        2400.0   
47                  0.500000                 7              0        4200.0   
54                  0.000000                 0              8        1000.0   
55                  0.000000                 0              6        6000.0   
...                      ...               ...            ...           ...   
8919                0.000000                 0              2        1500.0   
8929                0.166667                 5              0        1500.0   
8935                0.000000                 0              6        1500.0   
8944                0.000000                 0              2        4000.0   
8946                0.000000                 0              6        1000.0   

        PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
3       0.000000               NaN               0.0      12  
45      0.000000               NaN               0.0      12  
47      0.000000               NaN               0.0      12  
54    417.016763               NaN               0.0      12  
55      0.000000               NaN               0.0      12  
...          ...               ...               ...     ...  
8919    0.000000               NaN          

In [60]:
#Utilizando o comando fillna para trocar os valores nulos por zero
df = df.fillna(0)

In [61]:
#Verificando se ocorreu tudo bem com nosso fillna
df.isnull().sum()

CUST_ID                             0
BALANCE                             0
BALANCE_FREQUENCY                   0
PURCHASES                           0
ONEOFF_PURCHASES                    0
INSTALLMENTS_PURCHASES              0
CASH_ADVANCE                        0
PURCHASES_FREQUENCY                 0
ONEOFF_PURCHASES_FREQUENCY          0
PURCHASES_INSTALLMENTS_FREQUENCY    0
CASH_ADVANCE_FREQUENCY              0
CASH_ADVANCE_TRX                    0
PURCHASES_TRX                       0
CREDIT_LIMIT                        0
PAYMENTS                            0
MINIMUM_PAYMENTS                    0
PRC_FULL_PAYMENT                    0
TENURE                              0
dtype: int64

Também optei por retirar a parte C dos nosso IDs para não atrapalhar em análises futuras como na análise descritiva, e também renomeei as colunas com letras minúsculas. 

In [62]:
#Retirando o C do inicio do ID para transformar ele.
df['CUST_ID'] = df['CUST_ID'].str.replace('C', '')
df

CUST_ID      BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0      10001    40.900749           0.818182      95.40              0.00   
1      10002  3202.467416           0.909091       0.00              0.00   
2      10003  2495.148862           1.000000     773.17            773.17   
3      10004  1666.670542           0.636364    1499.00           1499.00   
4      10005   817.714335           1.000000      16.00             16.00   
...      ...          ...                ...        ...               ...   
8945   19186    28.493517           1.000000     291.12              0.00   
8946   19187    19.183215           1.000000     300.00              0.00   
8947   19188    23.398673           0.833333     144.40              0.00   
8948   19189    13.457564           0.833333       0.00              0.00   
8949   19190   372.708075           0.666667    1093.25           1093.25   

      INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                      95.40      0.000000             0.166667   
1                       0.00   6442.945483             0.000000   
2                       0.00      0.000000             1.000000   
3                       0.00    205.788017             0.083333   
4                       0.00      0.000000             0.083333   
...                      ...           ...                  ...   
8945                  291.12      0.000000             1.000000   
8946                  300.00      0.000000             1.000000   
8947                  144.40      0.000000             0.833333   
8948                    0.00     36.558778             0.000000   
8949                    0.00    127.040008             0.666667   

      ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                       0.000000                          0.083333   
1                       0.000000                          0.000000   
2                       1.000000                          0.000000   
3                       0.083333                          0.000000   
4                       0.083333                          0.000000   
...                          ...                               ...   
8945                    0.000000                          0.833333   
8946                    0.000000                          0.833333   
8947                    0.000000                          0.666667   
8948                    0.000000                          0.000000   
8949                    0.666667                          0.000000   

      CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                   0.000000                 0              2        1000.0   
1                   0.250000                 4              0        7000.0   
2                   0.000000                 0             12        7500.0   
3                   0.083333                 1              1        7500.0   
4                   0.000000                 0              1        1200.0   
...                      ...               ...            ...           ...   
8945                0.000000                 0              6        1000.0   
8946                0.000000                 0              6        1000.0   
8947                0.000000                 0              5        1000.0   
8948                0.166667                 2              0         500.0   
8949                0.333333                 2             23        1200.0   

         PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
0      201.802084        139.509787          0.000000      12  
1     4103.032597       1072.340217          0.222222      12  
2      622.066742        627.284787          0.000000      12  
3        0.000000          0.000000          0.000000      12  
4      678.334763        244.791237          0.000000      12  
...           ...               ...               ...     ...  
8945   325.594462         48.886365  

In [64]:
#Fazendo alteração do tipo da coluna de object para int

df['CUST_ID'] = df['CUST_ID'].astype(int)

In [65]:
#Renomeando as colunas de maíusculo para minúsculo para não termos problemas com o Postgres
df = df.rename(columns={'CUST_ID':'cust_id',
                       'BALANCE':'balance',
                       'BALANCE_FREQUENCY':'balance_frequency',
                       'PURCHASES':'purchases',
                       'ONEOFF_PURCHASES':'oneoff_purchases',
                       'INSTALLMENTS_PURCHASES':'installments_purchases',
                       'CASH_ADVANCE':'cash_advance',
                       'PURCHASES_FREQUENCY':'purchases_frequency',
                       'ONEOFF_PURCHASES_FREQUENCY':'oneoff_purchases_frequency',
                       'PURCHASES_INSTALLMENTS_FREQUENCY':'purchases_installments_frequency',
                       'CASH_ADVANCE_FREQUENCY':'cash_advance_frequency',
                       'CASH_ADVANCE_TRX':'cash_advance_trx',
                       'PURCHASES_TRX':'purchases_trx',
                       'CREDIT_LIMIT':'credit_limit',
                       'PAYMENTS':'payments',
                       'MINIMUM_PAYMENTS':'minimum_payments',
                       'PRC_FULL_PAYMENT':'prc_full_payment',
                       'TENURE':'tenure'})
df

cust_id      balance  balance_frequency  purchases  oneoff_purchases  \
0       10001    40.900749           0.818182      95.40              0.00   
1       10002  3202.467416           0.909091       0.00              0.00   
2       10003  2495.148862           1.000000     773.17            773.17   
3       10004  1666.670542           0.636364    1499.00           1499.00   
4       10005   817.714335           1.000000      16.00             16.00   
...       ...          ...                ...        ...               ...   
8945    19186    28.493517           1.000000     291.12              0.00   
8946    19187    19.183215           1.000000     300.00              0.00   
8947    19188    23.398673           0.833333     144.40              0.00   
8948    19189    13.457564           0.833333       0.00              0.00   
8949    19190   372.708075           0.666667    1093.25           1093.25   

      installments_purchases  cash_advance  purchases_frequency  \
0                      95.40      0.000000             0.166667   
1                       0.00   6442.945483             0.000000   
2                       0.00      0.000000             1.000000   
3                       0.00    205.788017             0.083333   
4                       0.00      0.000000             0.083333   
...                      ...           ...                  ...   
8945                  291.12      0.000000             1.000000   
8946                  300.00      0.000000             1.000000   
8947                  144.40      0.000000             0.833333   
8948                    0.00     36.558778             0.000000   
8949                    0.00    127.040008             0.666667   

      oneoff_purchases_frequency  purchases_installments_frequency  \
0                       0.000000                          0.083333   
1                       0.000000                          0.000000   
2                       1.000000                          0.000000   
3                       0.083333                          0.000000   
4                       0.083333                          0.000000   
...                          ...                               ...   
8945                    0.000000                          0.833333   
8946                    0.000000                          0.833333   
8947                    0.000000                          0.666667   
8948                    0.000000                          0.000000   
8949                    0.666667                          0.000000   

      cash_advance_frequency  cash_advance_trx  purchases_trx  credit_limit  \
0                   0.000000                 0              2        1000.0   
1                   0.250000                 4              0        7000.0   
2                   0.000000                 0             12        7500.0   
3                   0.083333                 1              1        7500.0   
4                   0.000000                 0              1        1200.0   
...                      ...               ...            ...           ...   
8945                0.000000                 0              6        1000.0   
8946                0.000000                 0              6        1000.0   
8947                0.000000                 0              5        1000.0   
8948                0.166667                 2              0         500.0   
8949                0.333333                 2             23        1200.0   

         payments  minimum_payments  prc_full_payment  tenure  
0      201.802084        139.509787          0.000000      12  
1     4103.032597       1072.340217          0.222222      12  
2      622.066742        627.284787          0.000000      12  
3        0.000000          0.000000          0.000000      12  
4      678.334763        244.791237          0.000000      12  
...           ...               ...               ...     ...  
8945   325.594462         

**2. Conexão com o Postgres**

Iniciaremos a conexão do pandas com o Postgree, para posteriormente criar um banco de dados chamaado 'cluster_clients' depois que inserirmos o dataframe pandas no Postgree.
Para isso, precisamos de algumas informações:

- Host: O endereço do servidor onde o PostgreSQL está rodando.

- Database: O nome do banco de dados ao qual você quer se conectar.
- User: O nome do usuário que tem acesso ao banco de dados.
- Password: A senha desse usuário.
- Port: A porta na qual o PostgreSQL está rodando (o padrão é 5432).

In [78]:
# Criando a conexão

pgconn = pg2.connect(
  host='localhost',
  user='seu_usuario',
  password='sua_senha',
  database='nome_do_database'
)

In [79]:
#Criando um cursor para auxiliar nas consultas e criação do database
pgcursor = pgconn.cursor()

In [80]:
#criando autocommit, isso facilita que não precisamos commitar automaticamente.

from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT 
pgconn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)

In [81]:
#desconectando todas as conexões que criamos

pgcursor.execute("""
SELECT pg_terminate_backend(pg_stat_activity.pid)
FROM pg_stat_activity
WHERE pg_stat_activity.datname = 'db' -- Replace 'db' with your database name
  AND pid <> pg_backend_pid();
""")

pgconn.commit()

Desconexão e criação do DataBase

In [70]:
#Verificando se existe nosso database

pgcursor.execute("SELECT 1 FROM pg_database WHERE datname = 'clusterclients'")
existe = pgcursor.fetchone()

#e se ele não identificar ele precisa criar

if not existe:
    pgcursor.execute('CREATE DATABASE clusterclients')

In [71]:
#commitando para salver todas as mudanças

pgconn.commit()

#fechando

pgconn.close()


In [82]:
#verificando o banco que está sendo usado
pgcursor.execute('SELECT current_database()')

In [83]:

#verificando o banco que está sendo usado
pgcursor.execute('SELECT current_database()')

#Criando o mecanismo de conexão
variable= 'postgresql+psycopg2://seu_usuario:sua_senha@localhost/clusterclients'
engine= create_engine(variable)

In [84]:
#Descrevendo nosso DataFrame df em uma tabela de um banco de dados SQL usando SQLAlchemy
df.to_sql('historical_purchases', engine, if_exists='replace', index=False)

950